File: Ejemplos/aula-java/src/main/java/es/joseluisgs/models/Alumno.java
package es.joseluisgs.models;

public class Alumno {
    private static int contador = 0;
    private final int id = nextId();
    private String nombre;
    private double nota;

    public Alumno(String nombre, double nota) {
        this.nombre = nombre;
        this.nota = nota;
    }

    private static int nextId() {
        return ++contador;
    }

    public static void resetId() {
        contador = 0;
    }

    public static Alumno create(String nombre, double nota) {
        return new Alumno(nombre, nota);
    }

    public static void respirar() {
        System.out.println("Respirando");
    }

    public static Alumno random() {
        return new Alumno("Alumno " + nextId(), Math.random() * 10);
    }

    public static int getContador() {
        return contador;
    }

    public int getId() {
        return id;
    }

    public String getNombre() {
        return nombre;
    }

//    public double getNota() {
//        return nota;
//    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public void setNota(double nota) {
        this.nota = nota;
    }

    public double getCalificacion() {
        return (int) (nota * 100) / 100.0;
    }

    public boolean isAprobado() {
        return nota >= 5;
    }

    public String getCalificacionDesc() {
        if (getCalificacion() < 5) {
            return "Suspenso";
        } else if (nota < 7) {
            return "Aprobado";
        } else if (nota < 9) {
            return "Notable";
        } else if (nota <= 10) {
            return "Sobresaliente";
        } else {
            return "Calificación no válida";
        }
    }

    @Override
    public String toString() {
        return "Alumno{" +
                "id=" + id +
                ", nombre='" + nombre + '\'' +
                ", calificacion=" + getCalificacion() +
                ", isAprobado=" + isAprobado() +
                ", calificacionDesc=" + getCalificacionDesc() +
                '}';
    }
}

File: Ejemplos/ProgramacionOrientadaObjetos/src/main/kotlin/aula/models/Alumno.kt
package aula.models

data class Alumno(val nombre: String, var nota: Double) {
    val califcacion get() = (this.nota * 100).toInt() / 100.0
    val isAprobado get() = califcacion >= 5
    val calificacionDesc
        get() = when (califcacion) {
            in 0.0..4.99 -> "Suspenso"
            in 5.0..6.99 -> "Aprobado"
            in 7.0..8.99 -> "Notable"
            in 9.0..10.0 -> "Sobresaliente"
            else -> "Calificación no válida"
        }
    val id = nextId()

    companion object {
        private var contador = 0
        private fun nextId() = ++contador
        fun reset() {
            contador = 0
        }

        fun create(nombre: String, nota: Double) = Alumno(nombre, nota)
        fun respirar() = println("Respirando...")

        // Uno de los usos es para crear objetos a partir de datos aleatorios...
        fun random() = Alumno("Alumno $contador", (0..10).random().toDouble())
    }

    override fun toString(): String {
        return "Alumno(id=$id, nombre='$nombre', califcacion=$califcacion, isAprobado=$isAprobado, calificacionDesc='$calificacionDesc')"
    }
}